### PR TITLE
Unlifted Newtypes

### PR DESCRIPTION
**Pull request by andrewthad**
This is a proposal for an `UnboxedNewtypes` extension. Most of the proposal focuses on motivating factors, since it is not immediately obvious why this is actually useful. I'm happy to receive any criticism dealing with underspecification of the extension. Also, if the motivations are unclear, let me know. I'm having a slightly difficult time articulating them without dumping GHC Core into the proposal, and I would like to make sure that others are able to make sense of them. Thanks for the consideration.

[rendered link](https://github.com/andrewthad/ghc-proposals/blob/unboxed_newtypes/proposals/0000-unlifted-newtypes.rst)

### CODE REVIEW

**Comment by treeowl** [comment]
I think it's perfectly obvious that we want this and also pretty easy to see why. +1

**Comment by goldfirere** [comment]
I am in support of this.

- I would encourage you to fill out the remaining sections of the proposal. The "proposed change specification" is easy, I think, but it should be spelled out. Other sections may be a bit harder. For example, how does this interact with #35? Specifically, having unlifted newtypes makes unlifted types more prevalent and harder to spot (there might be no `#`s in sight). This could cause unexpected strictness.

- I believe the implementation should be easy, so let's not worry about that. If no one else wants to, I'll do it.

**Comment by Icelandjack** [comment]
I'm +1 for a while, phantom types (`Ptr# a`) are a good use case. Mild concern: Before this `TYPE IntRep` only had a single inhabitant and could be processed by a closed type family, now presumably we would need an open type family to handle `PersonId`, `Bool#` ..?

----

This lets us define `Bool#` and pattern synonyms

```haskell
newtype Bool# :: TYPE IntRep where
  Bool# :: Int# -> Bool#

pattern False# :: Bool#
pattern False# = Bool 0#

pattern True# :: Bool#
pattern True# = Bool 1#
```

so that functions `(==#)`, `(==##)` return a `Bool#` rather than `Int#`:

```haskell
(==#)  :: Int#    -> Int#    -> Bool#
(==##) :: Double# -> Double# -> Bool#
```

and maybe 

```haskell
newtype Maybe# :: Type -> TYPE (SumRep '[LiftedRep, LiftedRep]) where
  Maybe# :: (# a | () #) -> Maybe# a

pattern Nothing# :: Maybe# a
pattern Nothing# = Maybe# (# | () #)

pattern Just# :: a -> Maybe# a
pattern Just# a = Maybe# (# a | #)
```

**Comment by goldfirere** [comment]
I would love it if @Icelandjack 's examples could be automatically generated somehow, not unlike how `UNPACK` can turn an `!Int` into an `Int#`:

```
data !Bool# = False# | True#
  -- note the bang
```

desugars to

```
newtype Bool# = MkBool# (# (# #) | (# #) #)
```

This goes well beyond the current proposal, which would lay good groundwork for future fun stuff to come.

**Comment by andrewthad** [comment]
I've added some more sections to the proposal:

- @goldfirere I need to think a little more about https://github.com/goldfirere/ghc-proposals/blob/unbanged-strict-patterns/proposals/0000-unbanged-strict-patterns.rst. I'll try to address that soon.
- @Icelandjack The opening of the unlifted kinds is a little strange, but to my understanding, they weren't really ever closed. It's not like we've ever had something like this sitting around: `data IntRepType = Int# | ...`. So it's although these kinds were closed in spirit, they weren't really closed. But yes, this could technically cause a closed type family that someone thought was total to not actually be total. I'll mention that in the drawbacks in the proposal.
- I just realized that I need to rename this to `UnliftedNewtypes`. That's a more accurate name, and arrays don't like feeling left out.

**Comment by Icelandjack** [comment]
What is then the kind of `newtype Id a = Id a`? Could it be

```haskell
newtype Id :: TYPE rep -> TYPE rep where
  Id :: a -> Id a
```

**Comment by Icelandjack** [comment]
```haskell
newtype Interval = Interval (# Int#, Int #)
```

should probably be `(# Int#, Int# #)`

**Comment by andrewthad** [comment]
@Icelandjack My intuition says that the `Id` newtype should be rejected since we still need to reject levity-polymorphic binders. Technically, it might be able to be safe since the `Id` data constructor is just a cast.

**Comment by Icelandjack** [comment]
Does that mean defining a `newtype` over `(# , #)` would not work?

```haskell
newtype Pair# :: TYPE rep -> TYPE rep' -> TYPE (TupleRep '[rep, rep']) where
  Pair# :: (# a, b #) -> Pair# a b
```

**Comment by andrewthad** [comment]
If `Id` doesn't work, then `Pair#` definitely work work. It has the same problem with creating a function that has a levity-polymorphic binder. But maybe neither of them actually have this problem. Let's see what happens currently if we try to make a levity-polymorphic `id` function:

```
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE MagicHash #-}

module LevityId where

import GHC.Types
import GHC.Prim

levityPolymorphicId :: forall (a :: TYPE r). a -> a
levityPolymorphicId = unsafeCoerce#
```

With `ghc-8.2.2`, we get this:

```
ghc: panic! (the 'impossible' happened)
  (GHC version 8.2.2 for x86_64-unknown-linux):
        runtimeRepPrimRep
  typePrimRep (a_aRu :: TYPE r_aRt)
  r_aRt
  Call stack:
      CallStack (from HasCallStack):
        prettyCurrentCallStack, called at compiler/utils/Outputable.hs:1133:58 in ghc:Outputable
        callStackDoc, called at compiler/utils/Outputable.hs:1137:37 in ghc:Outputable
        pprPanic, called at compiler/simplStg/RepType.hs:360:5 in ghc:RepType
```

That's not good. And I think we need something like that work for polymorphic unlifted newtypes to work. With `Id`, the type and hidden implementation of the data constructor would need to be:

    idConstructor :: forall (a :: TYPE r). a -> Id a
    idConstructor = unsafeCoerce#

Sorry this has been a bit of a ramble. I think that technically, polymorphic unlifted newtypes should be possible. It doesn't actually seem to break the levity-polymorphic binder rule.

**Comment by goldfirere** [comment]
That `levityPolymorphicId` should compile, I think. I've reported the panic as [#14561)[https://ghc.haskell.org/trac/ghc/ticket/14561]. In any case, I think the `Id` newtype above is conceivable. There may be some challenges in accepting that -- and it's actually unclear whether that case is covered by this proposal -- but I do think that GHC Core is capable of expressing such a thing.

**Comment by andrewthad** [comment]
I've made two additions to the proposal:

1. Specify that levity-polymorphic type variables can be used in an unlifted newtype.
2. Clarify how the strictness of pattern bindings should work.

Also, I've renamed the proposal to Unlifted Newtypes.

**Comment by andrewthad** [comment]
Other random asides. I like @Icelandjack's list of things that could be done with this. I had not considered the possibility of a `Maybe#`, but a variant of it seems to provide a compelling argument for the inclusion of levity polymorphism in unlifted newtypes:

    newtype Maybe# :: (a :: TYPE (r :: RuntimeRep)) -> TYPE (SumRep '[r, TupleRep '[]]) where
      Maybe# :: (# a | (# #) #) -> Maybe# a

It's neat to have this as a reusable type (even if its not possible write super reusable functions over it).

Also @goldfirere, thanks for opening up that issue. After reading the [Mega-hack for coerce](https://github.com/ghc/ghc/blob/e7985ed23ddc68b6a2e4af753578dc1d9e8ab4c9/libraries/base/Unsafe/Coerce.hs#L40-L55), I convinced myself that the compulsory unfolding would forbid this, but it's good to know that it should be doable. I've wanted a kind-safe variant of `unsafeCoerce#` for a while, and it's neat knowing that it might just be possible.

**Comment by goldfirere** [comment]
The examples in this thread have included lots of explicit kind annotations. I don't think these will be necessary in practice (but they are helpful for exposition) -- GHC should have no trouble inferring the kind of the newtype from looking at the kind of its one argument.

**Comment by andrewthad** [comment]
Agreed. They should be able to be inferred. I'll clarify that in the proposal.

**Comment by andrewthad** [comment]
Another wrinkle I just thought about. The following is allowed:

    newtype Foo = Foo Bar
    newtype Bar = Bar Foo

At compile time, this is fine. But, in the presence of `UnliftedNewtypes`, what is the kind of `Foo`? We probably want it to just be `TYPE 'LiftedRep`. Even worse:

    newtype Foo = Foo (# Int#, Bar #)
    newtype Bar = Bar Foo

The kind of `Foo` is infinite, which should be a compile-time error, but some care will need to be taken to prevent this from hanging compilation.

**Comment by goldfirere** [comment]
Er, good point. I hadn't thought about recursion. Silly me. Inference may bit hard in that case. We could fairly easily handle the non-recursive case, and then require an annotation if there is recursion...

**Comment by andrewthad** [comment]
I think that  it might be preferable to prohibit recursive unlifted newtypes. I don’t think they will ever actually work.

**Comment by treeowl** [comment]
How does this interact with data families? Will we now support data families with unlifted result kinds? These would support only `newtype instance`, not `data instance`.

```haskell
data family Foo (n :: Nat) :: TYPE ('TupleRep '[ 'IntRep, 'LiftedRep])
newtype instance Foo 0 = X (# Int#, Bool #)
newtype instance Bar 1 = Y (# Int#, Maybe Bool #)
```

**Comment by andrewthad** [comment]
@treeowl Good point. I had not considered that interaction. Currently, even attempting to define such a data family errors with:

```
unlifted_data_family.hs:8:1: error:
    • Kind signature on data type declaration has non-* return kind
        TYPE 'IntRep
    • In the data family declaration for ‘Foo’
  |
8 | data family Foo (a :: Type) :: TYPE 'IntRep
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

They should certainly be supported though (especially since type families allow unlifted return kinds). I'll update the proposal to reflect this.

**Comment by andrewthad** [comment]
Updates that have been made:

- Explicitly require unlifted data families to be supported
- Specify how recursive newtypes should work. I recommend reading this part, but others may disagree with what is proposed here.

**Comment by andrewthad** [comment]
Oh, and I also added some discussion of the interaction with GADT syntax.  Thanks for all the feedback on this so far! It has been helpful for identifying the underspecified areas of the proposal.

**Comment by andrewthad** [comment]
This proposal also interacts with Coercible in the obvious way.  I’m putting this here mostly as a reminder to myself to mention this in the proposal.

**Comment by simonpj** [comment]
I'm not against this, though I'm not wholly convinced of the power-to-weight ratio.

Some thoughts

`.  Under "specification" you say "The calling convention for the IdC# data constructor does not depend on r".  But it certainly does!  Today, for each newtype GHC defines a data constructor wrapper, thus
```
newtype T a = MkT (a->a)
===>
MkT :: forall a. (a->a) -> T a
MkT = /\a. \(x :: a->a). x |> co
```
where `co` is a suitable coercion.   But for `IdC` the lambda would bind a levity-polymorphic variable.

You could insist that `IdC` appears only saturated.  Or something.  But you need something. c.f. Trac #14561.

2. Don't forget to talk about what happens with `Coercible`.   That has to become levity-polymorphic too, I guess. I wonder if other classes have to as well?

3.  There's an impact on Core that I would like to see spelled out:.  In particular, the equality `~R#` has to become levity polymorphic and take a levity argument.  See `Note [The equality types story]` in `TysPrim`.

4. The story around recursion (which is stated a bit imprecisely) is easier to understand if you do kind inference.   In your example with `Sneak`, the type constructors would get infinite kinds, like
```
Sneak :: Type (TupleRep [IntRep, TupleRep [IntRep, ...]])
```
and just as terms can't have infinite types, so type can't have infinite kinds.

5. I was unconvinced by Motivation: Unlfited arrays as a library, because it seems to amount to adding a phantom type parameter to `ByteArray#`.  You need the primops anyway; so we could do this whole section without changing newtypes, just by giving primitive `ByteArray#` (or a newly named `UnliftedArray#` a phantom parameter.

**Comment by Icelandjack** [comment]
Here is a Trac ticket for Simon's *2.*: [`#13595`: Should ‘coerce’ be levity polymorphic?](https://ghc.haskell.org/trac/ghc/ticket/13595)

**Comment by andrewthad** [comment]
@simonpj Responding to your thoughts:

1. Yes. The proposal is currently totally wrong when it says that "The calling convention for the IdC# data constructor does not depend on r". I had originally assumed that `unsafeCoerce` somehow magically subvert the normal levity polymorphic binder restriction, but now I see that it doesn't actually work that way. The outcome of [Trac #14561](https://ghc.haskell.org/trac/ghc/ticket/14561) will determine whether or not newtypes can have polymorphic type variables. Quoting the three options presented so far on that ticket:
    i. Do not allow unsafeCoerce on unboxed values. I don't know how much it is used. (SPJ)
    ii. Check that all uses are saturated. Perhaps in the desugarer. (SPJ)
    iii. Allow silly uses of unsafeCoerce# to cause a panic.
It's worth noting that the only one of these that would make levity polymorphic unlifted newtypes both possible and practical would be the second one. I'll work on an alternative section in the proposal that specifies not allowing levity-polymorphic unlifted newtypes, but my preference is to allow them.
2. I've added a brief discussion of `Coercible` to the proposal.
3. The impact on core should be mentioned in the proposal. I need to do some reading on this since I don't understand `~R#` very well myself, but I'll try to get something specified soon.
4. I've updated the section on recursion to explicitly show the expansion you mention.
5. You're right. That one could certainly be solved by just adding a phantom type parameter to `ArrayArray#`. But, then what if someone wanted the old `ArrayArray#` interface back? We could recreate that type with an unlifted lifted around the new one. I consider this the least compelling example though because the existing `ArrayArray#` interface is really only untyped as a result of it preceeding levity polymorphism, and it should really just be fixed as an issue of its own.

I want to try to address to power-to-weight ratio. This proposal is really only helpful under pretty narrow circumstances. Let's consider two definitions of `PersonId`, a function that operates on it, and an application of the function:

    newtype PersonId = PersonId Word
    newtype PersonId = PersonId Word#
    
    -- Assume this is something large that won't be inlined
    nextPersonId :: PersonId -> PersonId

    bigFunction :: Foo -> Bar
    bigFunction foo =
      ...
      let jane = nextPersonId jordan
      ...

Regardless of which `PersonId` we are using, we're going to end up avoiding boxing when we call `nextPersonId` inside of `bigFunction`. If we use the unlifted `PersonId`, then obviously there cannot be any boxing, but if we use the lifted one, then work-wrapper (plus inlining of the wrapper) is still going to save the day. So, there's no difference here. It only matters when bigFunction is higher-order:

    bigFunction :: (PersonId -> PersonId) -> Foo -> Bar
    bigFunction f foo =
      ...
      let jane = f jordan
      ...

Now the performance of `bigFunction nextPersonId` will depend on which definition of `PersonId` and `nextPersonId` we used. This has been verified in [this gist](https://gist.github.com/andrewthad/1e6b6d7c104b02a25f367e341bd2a2da).

So for this to really matter for some type `Foo`, you have to have two things:

1. A higher-order function that takes a function operating on `Foo`
2. The need to hide `Foo`'s data constructor. Either `Foo` has an invariant preserved by functions that operate on it, or you don't want people to create them out of thin air.

Where does this happen? In the present, it's in APIs that do CPS to control to lifetimes of pointers or pointer-like resources. In the future, I think the need for this will show up in libraries that use linear types. Most applications of linear types that I've seen rely heavily on both CPS and data-constructor hiding to preserve their invariants, so this could help reduce allocations there as well.

**Comment by andrewthad** [comment]
@simonpj Also, is the code used in section 5.1.3 of the Linear Types paper available? If the internal implementation is what I think it is, I might be able demonstrate a very specific way that unlifted newtypes would be helpful.

**Comment by andrewthad** [comment]
I found the code I was looking for. It's in the [linear-types](https://github.com/tweag/linear-types) repo on github. From section 5.1.3 of the Linear Types paper:

> Using this approach, we were able to ensure by construction that the "linear/packed" implementations in Fig. 7 were completely non-allocating, rather than depending on the optimiser. This results in better performance for the linear, compared to monadic version of the serialised-data transformations.

For context, the function that has been levity polymorphized is `caseTree`:

```
caseTree :: forall (rep :: RuntimeRep) (res :: TYPE rep) b.
             Packed (Tree ': b) ⊸  (Packed (Int ': b) ⊸ res ) -> (Packed (Tree ': Tree ': b) ⊸ res ) -> res
```

And the place elsewhere in the code that can take advantage of this is the definition of `mapLeaves`. From the paper, here is the allocating (less performant) of `mapLeaves`:

```
mapLeaves :: (Int → Int) → Packed ′[Tree] ⊸ Packed ′[Tree]
mapLeaves fn pt = newBuffer (extract ◦ go pt)
    where
    extract (inp, outp) = case done inp of () → finish outp
    go :: Packed (Tree ′: r) ⊸ Needs (Tree ′: r) t ⊸ (Packed r, Needs r t)
    go p = caseTree p
      (λp o → let (x, p’) = read p in (p’, writeLeaf (fn x) o))
      (λp o → let (p’, o’) = go p (writeBranch o) in go p’ o’)
```

It is necessarily allocating because of several factors that work together:

1. `go` is not tail recursive
2. `go` returns a lifted tuple
3. The tuple `go` returns contains lifted types.

Since we cannot change (1), we have to change both (2) and (3) to prevent allocations from happening. It's easy to fix (2) by just using an unboxed tuple instead. To fix (3), we have to change the internal representation of `Packed` and `Needs`. From the repository, we can see how this is done. The definitions of [Has#](https://github.com/tweag/linear-types/blob/84d4a2c0fec6b6d86aaa93340da42d0fe315ec07/Examples/src/Cursors/UnboxedHas.hs#L35), which is what `Packed` has been renamed to, and [Needs](https://github.com/tweag/linear-types/blob/84d4a2c0fec6b6d86aaa93340da42d0fe315ec07/Examples/src/Cursors/Mutable.hs#L69) are:

    type Has# (a :: [Type]) = Addr#
    type Needs (l :: [Type]) t = (# Int#, Addr# #)

But although this implementation is faster, it is no longer safe. It is not possible to hide the internal representation. Unlifted newtypes would solve this because then you could hide the internal representation.

As pure speculation, I'll offer this: I suspect that a lot of other zero-allocation applications of linear types are going to need unlifted newtypes. Honestly, this is the main reason I that I actually want this extension. I didn't think it would be appropriate to mention this in the proposal since linear types have not been approved yet and because it is, like I said, very speculative.

**Comment by goldfirere** [comment]
I disagree with Simon's point (3): `~R#` is already fully kind-polymorphic: `(~R#) :: forall k1 k2. k1 -> k2 -> Type`. `Coercible` is also fully kind-polymorphic (though homogeneous). The only change of this sort that would have to happen is to make `coerce` levity-polymorphic. Currently, we have `coerce :: forall (a :: Type) (b :: Type). Coercible @Type a b => a -> b`. We would need `coerce :: forall (r :: RuntimeRep) (a :: TYPE r) (b :: TYPE r). Coercible @(TYPE r) a b => a -> b`. (Note that it is levity-homogeneous because newtypes will have the same representations as their unfoldings.)

Bottom line: I don't think Core is touched by this proposal.

**Comment by goldfirere** [comment]
- Reading the updates to the proposal, I disagree with the suggestion that uninhabited newtypes should be inferred to have levity-polymorphic kinds. GHC currently *never* infers levity-polymorphism, and I think that's a nice design principle to keep. If someone writes `newtype Void = Void Void`, we don't want them getting obscure levity-polymorphism errors later.

- The proposal seems a bit confused between GND and `coerce`. It says that GND works with unlifted newtypes as a straightforward extension of how it works now, but then states that `coerce` would *not* be generalized in the proposal. But GND uses `coerce`! So I think these statements are in conflict.

- As I said above, `Coercible` does not need to be generalized (in contrast to the current text of the proposal). It's `coerce` that needs to be generalized.

- Your `Foo` data family instances are ill-kinded: the second one does not have an `IntRep` representation. Do you mean the return kind of the data family to itself be the result of a type family application? Such a thing would be sensible, but it may be beyond GHC's ability to deal with type families in representations (an implementation limitation we currently have).

Thanks for continuing to work on this proposal! This is already quite a bit more involved than I thought it would be.

**Comment by andrewthad** [comment]
@goldfirere I agree with you on all of the points you raised. It is certainly more consistent with the design of everything else to not infer levity polymorphism for uninhabited newtypes. I had forgotten that GND needs `coerce`, so I've revised discussion of `Coercible` to require a levity-polymorphic `coerce`. Also, I had miswritten `Foo`, so I've changed that.

**Comment by andrewthad** [comment]
After neglecting to do so for two months, I've made a few more tweaks to the proposal. I would like submit it to the committee for review.

**Comment by andrewthad** [comment]
@rrnewton Glad to hear that one of the Linear Types paper's authors felt this shortcoming as well. Hopefully this extension can plug that gap.

**Comment by treeowl** [comment]
By the way, I came up with a fairly strong argument in favor of levity-polymorphic newtypes. Suppose we define the very-polymorphic continuation monad transformer:

```haskell
newtype ContT (r :: k) (m :: k -> TYPE rep) a = ContT {runContT :: (a -> m r) -> m r}
```

Now we'd like to be able to define the vanilla continuation monad in terms of it. For the standard continuation monad, we can just write

```haskell
type Cont r = ContT r Identity
```

But because `Identity` isn't (yet) levity-polymorphic, that can only give us

```haskell
Cont :: Type -> Type -> Type
```

With levity-polymorphic `Identity`, we win

```haskell
Cont :: TYPE rep -> Type -> Type
```

which is considerably more useful.
